### PR TITLE
Implement a TestClient class to test routes more efficiently

### PR DESCRIPTION
**Pull request by ElyasMag**
**Description**

This PR adds a new class TestClient that can be used to directly test methods.

To use, a robyn.testing.TestClient object must be created using a Robyn application as a parameter. HTTP method calls can be simulated using the functions get, post, delete, patch, options, head, trace and connect. Specifications of the input parameters and output object match those of the requests library.

Issue #507: https://github.com/sparckles/robyn/issues/507

### CODE REVIEW

File: robyn/testing.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+from robyn import Request, Response
+
+def get_route(app, path):
+	routes = app.router.get_routes()
+	r = None
+	for route in routes:
+		if route.route == path:
+			r = route
+			break
+	return r
```
Comment (sansyrox) [inline robyn/testing.py:20]: We can move this function inside the class only 😄

**Comment by sansyrox** [comment]
Hey @ElyasMag 👋 

Thanks for the PR 😄 I have had an initial look. I will have a deeper look as you make more progress 😄 

Do let me know if you have any questions from my end!

**Comment by ElyasMag** [comment]
> Hey @ElyasMag 👋
> 
> Thanks for the PR 😄 I have had an initial look. I will have a deeper look as you make more progress 😄
> 
> Do let me know if you have any questions from my end!

Hello! I'm new here. I'd appreciate any help I can get. I have one main question: If you look at my testing.py, you'll see that I'm passing None to the route handler. I tried to create a Request object to pass into it, but I get an error saying "class has no constructor." If you have any advice to offer, it would be much appreciated.

**Comment by ElyasMag** [comment]
> > Hey @ElyasMag 👋
> > Thanks for the PR 😄 I have had an initial look. I will have a deeper look as you make more progress 😄
> > Do let me know if you have any questions from my end!
> 
> Hello! I'm new here. I'd appreciate any help I can get. I have one main question: If you look at my testing.py, you'll see that I'm passing None to the route handler. I tried to create a Request object to pass into it, but I get an error saying "class has no constructor." If you have any advice to offer, it would be much appreciated.

@sansyrox Mentioning in case you didn't get a notification.

**Comment by sansyrox** [comment]
Hey @ElyasMag 👋 

>  I tried to create a Request object to pass into it, but I get an error saying "class has no constructor." If you have any advice to offer, it would be much appreciated.


There are two reasons for this 

One is that you need to inspect the kind of params the handler is expecting 
And second, where are you importing the request class from?

**Comment by ElyasMag** [comment]
@sansyrox Sorry, I can't seem to figure it out,
1. As far as I can tell, the only parameter passed into the handler - that is, the function written by the developer of a server using Robyn - is the Request object.
2. I'm importing the Request class from the robyn library, which I believe comes from src/types/request.rs & robyn/robyn.pyi

I thought I would have to create a class implementing the Request class, but Request is not an acceptable base type. Creating my own class with the same fields as the Request class would probably work, but doesn't sound like a reasonable solution. I can't figure out any other method of passing a Request object to the handler.
Once again, advice would be greatly appreciated.

File: robyn/testing.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+# the "app" parameter in the constructor refers to a Robyn application
+# app.router leads to the router, and from there we can use app.router.get_routes()
+# to retrieve the routes the server supports. We can search said list for methods
+# to match test requests made by the server's developer
+
+# it's important to understand the following classes, at least right now:
+# Route, Request, Response
+
+import asyncio
+from robyn import Request, Response
```
Comment (parth-gr) [inline robyn/testing.py:11]: @ElyasMag try this

from robyn.robyn import Router

router = Router(...)
print router.getRoute()

File: robyn/testing.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        if route == None:
+            return None
+        req = TestRequest(method = "POST")
+        if files != None:
+            body, header = encode_multipart_formdata(files)
+            req.headers["content-type"] = header
+            req.body = list(body)
+		return asyncio.run(route.function.handler(req))
+        
+
```
Comment (parth-gr) [inline robyn/testing.py:94]: only one line

**Comment by parth-gr** [comment]
Update the description of the PR and also attach the issue,

Undraft the PR so you can get a round of review from the maintainers

**Comment by sansyrox** [comment]
Hey @ElyasMag 👋 

> @sansyrox Sorry, I can't seem to figure it out, As far as I can tell, the only parameter passed into the handler - that is, the function written by the developer of a server using Robyn - is the Request object.
I'm importing the Request class from the robyn library, which I believe comes from src/types/request.rs & robyn/robyn.pyi
I thought I would have to create a class implementing the Request class, but Request is not an acceptable base type. Creating my own class with the same fields as the Request class would probably work, but doesn't sound like a reasonable solution. I can't figure out any other method of passing a Request object to the handler.
Once again, advice would be greatly appreciated.

Apologies, I missed this message :( .

Thanks @parth-gr for the help 💖 

Do you still have any queries??

**Comment by ElyasMag** [comment]
> Hey @ElyasMag 👋
> 
> > @sansyrox Sorry, I can't seem to figure it out, As far as I can tell, the only parameter passed into the handler - that is, the function written by the developer of a server using Robyn - is the Request object.
> > I'm importing the Request class from the robyn library, which I believe comes from src/types/request.rs & robyn/robyn.pyi
> > I thought I would have to create a class implementing the Request class, but Request is not an acceptable base type. Creating my own class with the same fields as the Request class would probably work, but doesn't sound like a reasonable solution. I can't figure out any other method of passing a Request object to the handler.
> > Once again, advice would be greatly appreciated.
> 
> Apologies, I missed this message :( .
> 
> Thanks @parth-gr for the help 💖
> 
> Do you still have any queries??
> 
> And for the future, if I don't reply to a query within 48 hours, feel free to reach out to me on Discord 😄

Hello @sansyrox!
I'll keep that in note for the future. The CodeDay Labs program is ending soon but I'd definitely like to continue working on this after.
How is my approach looking? I noticed that in the description of the issue, the functions' arguments and return values matched those of the request library instead of those of the robyn library, so I tried to implement them similarly.

File: testclient_example/testclient.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
+from testing import TestClient
+from robyn_server import app
```
Comment (parth-gr) [inline testclient_example/testclient.py:2]: myapp

File: robyn/testing.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+                    body = body + element[0] + "=" + element[1] + "&"
+                body = body[:-1]
+                req.body = list(body)
+                req.headers["content-type"] = "application/x-www-form-urlencoded"
+            req.headers["content-length"] = len(req.body)
+        elif json == None:
+            req.headers["content-type"] = "application/json"
+            req.body = list(dumps(json))
+            req.headers["content-length"] = len(req.body)
+        #run the handler function
```
Comment (parth-gr) [inline robyn/testing.py:147]: leave a line

File: robyn/testing.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+        response = asyncio.run(route.function.handler(req))
+        #turn the output into a requests.Response object
+        r = Response()
```
Comment (parth-gr) [inline robyn/testing.py:150]: attached the issue why not used the robyn response and also the reason as a todo

File: robyn/testing.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+        #input parameters
+        if params != None:
+            if type(params) == dict:
+                req.queries = params
+            elif type(params) == list:
+                for param in params:
+                    req.queries[param[0]] = param[1]
+        #input body
+        if files != None:
+            body, header = encode_multipart_formdata(files)
```
Comment (parth-gr) [inline robyn/testing.py:121]: can we call a separate function for this, single function not looks good

File: robyn/testing.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+            req.headers["content-length"] = len(req.body)
+        elif json == None:
+            req.headers["content-type"] = "application/json"
+            req.body = list(dumps(json))
+            req.headers["content-length"] = len(req.body)
+        #run the handler function
+        response = asyncio.run(route.function.handler(req))
+        #turn the output into a requests.Response object
+        r = Response()
+        r.status_code = response.status_code
```
Comment (parth-gr) [inline robyn/testing.py:151]: this should be a separate function

File: testclient_example/testclient.py [hunk 9]
```diff
@@ -0,0 +1,18 @@
+from robyn.testing import TestClient
```
Comment (parth-gr) [inline testclient_example/testclient.py:1]: We need to add a document step on the glow how to run this from the end user perspective

**Review by parth-gr — CHANGES_REQUESTED** [review]

File: robyn/testing.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+# uses a new TestRequest class instead of robyn.Request
+
+import asyncio
+from json import dumps
+from typing import Callable, Optional, Union
+from urllib3 import encode_multipart_formdata
+from robyn import HttpMethod
+from requests.models import Response
+
+# until we figure out a better method, I will be copying the classes over
```
Comment (parth-gr) [inline robyn/testing.py:12]: explain the issue OF CONSTRUCTOR

Why we need to do this provide URL links,

File: robyn/testing.py [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+    def connect(self, method_path, **kwargs):
+        return self.do_test_request(HttpMethod.CONNECT, method_path, **kwargs)
+    
+    def do_test_request(self, method, method_path, data=None, json=None, headers=None, files=None, params=None):
+        route = self.get_route(method_path, method)
+        if route == None:
+            return None
+        req = TestRequest(method = method)
+        #default headers
+        req.headers["host"] = "localhost"
```
Comment (parth-gr) [inline robyn/testing.py:103]: have a separate function for headers,
And do it for every param

File: robyn/testing.py [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+                for param in params:
+                    req.queries[param[0]] = param[1]
+        #input body
+        if files != None:
+            body, header = encode_multipart_formdata(files)
+            req.headers["content-type"] = header
+            req.body = list(body)
+            req.headers["content-length"] = len(req.body)
+        elif data != None:
+            if type(data) == bytes:
```
Comment (parth-gr) [inline robyn/testing.py:126]: I don't think we need to do for single type there might a function in python that would be handelling it

**Review by IdoKendo — CHANGES_REQUESTED** [review]
Hey @ElyasMag 👋 
Awesome work! I've left a few comments, let me know if anything's unclear 😃

File: testclient_example/testclient.py [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+    assert response.status_code == 200
+    assert response.text == "Hello, Robyn!"
+
+def test_upload():
+    file_data = b"This is a test file"
+    response = client.post("/upload", files={"file": ("test_file.txt", file_data)})
+    assert response.status_code == 200
+    assert response.json() == {"message": "File uploaded successfully"}
+
+test_index()
```
[the remark this answers is not in the document]
Comment (IdoKendo) [reply testclient_example/testclient.py:17]: When you run pytest it discovers functions that begin or end with test and runs them.

**Comment by ElyasMag** [comment]
@sansyrox
Hey! I have a couple of questions to ask, if you don't mind answering them!
- What is the Identity field of the Request class?
- Do redirects need to be handled by the testing client? I'm looking at the Digest authentication and it involves some back-and-forth which the HttpDigestAuth class attaches hooks to the request for. This involves a lot of extra functions so if that's the case, then I might need to revisit my approach for the test requests entirely.

File: robyn/testing.py [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+            req.body = list(dumps(json))
+            req.headers["content-length"] = len(req.body)
+    def add_input_parameters(self, req, params):
+        if params != None:
+            if type(params) == dict:
+                req.queries = params
+            elif type(params) == list:
+                for param in params:
+                    req.queries[param[0]] = param[1]
+    def add_cookies(self, req, cookies):
```
Comment (parth-gr) [inline robyn/testing.py:145]: leave a line between functons

**Comment by IdoKendo** [comment]
Hey @ElyasMag 👋 
I see that there are quite a few formatting/linting issues - We have pre-commit set up in the repo with `black` and `ruff` to avoid these and fix them at the commit step.
If you follow the [local development guide](https://github.com/sparckles/robyn#%EF%B8%8F-to-develop-locally) you can see how to set it up.
You can run `pre-commit run --all-files` after you set it up to run it on all files that you've previously committed.
Feel free to reach out if you're having issues!

**Pull request closed without merging**